File: ChessQuad/public/scripts/Piece.js
import { NormalMove, AttackMove } from "./Movement.js";
import Utils from "./Utilities.js";

const PieceType = Object.freeze({
  PAWN: {
    name: 'Pawn',
    symbol: 'P',
    value: 100
  },
  KNIGHT: {
    name: 'Knight',
    symbol: 'N',
    value: 300,
    legals: [-Utils.ORDER - 2, -Utils.ORDER * 2 - 1, -Utils.ORDER * 2 + 1, -Utils.ORDER + 2, Utils.ORDER - 2, Utils.ORDER * 2 - 1, Utils.ORDER * 2 + 1, Utils.ORDER + 2]
  },
  BISHOP: {
    name: 'Bishop',
    symbol: 'B',
    value: 300,
    legals: [-Utils.ORDER - 1, 1 - Utils.ORDER, Utils.ORDER - 1, Utils.ORDER + 1]
  },
  ROOK: {
    name: 'Rook',
    symbol: 'R',
    value: 500,
    legals: [-Utils.ORDER, -1, 1, Utils.ORDER]
  },
  GUARDIAN: {
    name: 'Guardian',
    symbol: 'G',
    value: 400,
    legals: [-Utils.ORDER - 1, -Utils.ORDER, 1 - Utils.ORDER, -1, 1, Utils.ORDER - 1, Utils.ORDER, Utils.ORDER + 1]
  },
  QUEEN: {
    name: 'Queen',
    symbol: 'Q',
    value: 900,
    legals: [-Utils.ORDER - 1, -Utils.ORDER, 1 - Utils.ORDER, -1, 1, Utils.ORDER - 1, Utils.ORDER, Utils.ORDER + 1]
  },
  KING: {
    name: 'King',
    symbol: 'K',
    value: Infinity,
    legals: [-Utils.ORDER - 1, -Utils.ORDER, 1 - Utils.ORDER, -1, 1, Utils.ORDER - 1, Utils.ORDER, Utils.ORDER + 1]
  }
});

export default class Piece {
  constructor(position, alliance, pieceType) {
    this.position = position;
    this.alliance = alliance;
    this.pieceType = pieceType;
    this.isFirstMove = true;
  }

  calculateLegalMoves (board) {
    return vectorMoves(board, this);
  }

  movePiece (move) {
    const clonedPiece = Utils.clone(this);
    clonedPiece.position = move.destinationIndex;
    clonedPiece.alliance = move.movedPiece.alliance;
    return clonedPiece;
  }

  equals (other) {
    if (this == other) {
      return true;
    } else if (other instanceof Piece) {
      return this.position == other.position
        && this.alliance == other.alliance
        && this.pieceType == other.pieceType
        && this.isFirstMove == other.isFirstMove;
    }
    return false;
  }

  toString () {
    return (this.alliance.direction < 0 ? 'W' : 'B') + this.pieceType.symbol;
  }
}

export class Pawn extends Piece {
  constructor(position, alliance) {
    super(position, alliance, PieceType.PAWN);
  }

  calculateLegalMoves (board) {
    const legalMoves = [], candidateDestination = {};

    candidateDestination.index = this.position + Utils.ORDER * this.alliance.direction;

    for (const adjacentAttackIndex of [-1, 1]) {
      candidateDestination.square = board.getSquare(candidateDestination.index + adjacentAttackIndex);
      if (candidateDestination.square.isInvalid) continue;
      if (candidateDestination.square.isOccupied) {
        const pieceAtDestination = candidateDestination.square.piece;
        const pieceAlliance = pieceAtDestination.alliance;

        if (pieceAlliance != this.alliance) {
          legalMoves.push(new AttackMove(board, this, pieceAtDestination));
        }
      }
    }

    candidateDestination.square = board.getSquare(candidateDestination.index);
    if (!candidateDestination.square.isOccupied) {
      legalMoves.push(new NormalMove(board, this, candidateDestination.index));

      if (this.isFirstMove) {
        candidateDestination.index += Utils.ORDER * this.alliance.direction;
        candidateDestination.square = board.getSquare(candidateDestination.index);
        if (!candidateDestination.square.isOccupied) {
          legalMoves.push(new NormalMove(board, this, candidateDestination.index));
        }
      }
    }

    return legalMoves;
  }
}
export class Knight extends Piece {
  constructor(position, alliance) {
    super(position, alliance, PieceType.KNIGHT);
  }

  calculateLegalMoves (board) {
    const legalMoves = [], candidateDestination = {};
    const self = this;
    for (const currentCandidate of self.pieceType.legals) {
      candidateDestination.index = currentCandidate + self.position;
      if (candidateDestination.index <= Utils.ORDER || candidateDestination.index >= Utils.TOT_SQR) continue;
      candidateDestination.square = board.getSquare(candidateDestination.index);

      if (candidateDestination.square.isInvalid) continue;

      if (candidateDestination.square.isOccupied) {
        const pieceAtDestination = candidateDestination.square.piece;
        const pieceAlliance = pieceAtDestination.alliance;

        if (pieceAlliance != self.alliance) {
          legalMoves.push(new AttackMove(board, self, pieceAtDestination));
        }
      } else {
        legalMoves.push(new NormalMove(board, self, candidateDestination.index));
      }
    }
    return legalMoves;
  }
}
export class Bishop extends Piece {
  constructor(position, alliance) {
    super(position, alliance, PieceType.BISHOP);
  }
}
export class Rook extends Piece {
  constructor(position, alliance) {
    super(position, alliance, PieceType.ROOK);
  }
}
export class Queen extends Piece {
  constructor(position, alliance) {
    super(position, alliance, PieceType.QUEEN);
  }
}
export class King extends Piece {
  constructor(position, alliance) {
    super(position, alliance, PieceType.KING);
  }

  calculateLegalMoves (board) {
    const legalMoves = [], candidateDestination = {};

    for (const currentCandidate of this.pieceType.legals) {
      candidateDestination.index = currentCandidate + this.position;
      candidateDestination.square = board.getSquare(candidateDestination.index);

      if (candidateDestination.square.isInvalid) continue;

      if (candidateDestination.square.isOccupied) {
        const pieceAtDestination = candidateDestination.square.piece;
        const pieceAlliance = pieceAtDestination.alliance;

        if (pieceAlliance != this.alliance) {
          legalMoves.push(new AttackMove(board, this, pieceAtDestination));
        }
      } else {
        legalMoves.push(new NormalMove(board, this, candidateDestination.index));
      }
    }
    return legalMoves;
  }
}

function vectorMoves (board, self) {
  const legalMoves = [], candidateDestination = {};

  for (const currentCandidate of self.pieceType.legals) {
    candidateDestination.index = self.position;

    while (true) {
      candidateDestination.index += currentCandidate;
      candidateDestination.square = board.getSquare(candidateDestination.index);

      if (candidateDestination.square.isInvalid) break;

      if (candidateDestination.square.isOccupied) {
        const pieceAtDestination = candidateDestination.square.piece;
        const pieceAlliance = pieceAtDestination.alliance;

        if (pieceAlliance != self.alliance) {
          legalMoves.push(new AttackMove(board, self, pieceAtDestination));
        }
        break;
      } else {
        legalMoves.push(new NormalMove(board, self, candidateDestination.index));
      }
    }
  }
  return legalMoves;
}

File: java-games/ChessEngine/src/chess/engine/pieces/Piece.java
package chess.engine.pieces;

import java.util.Collection;

import chess.engine.Alliance;
import chess.engine.board.Board;
import chess.engine.board.Move;

public abstract class Piece {
  protected final PieceType pieceType;
  protected final int piecePosition;
  protected final Alliance pieceAlliance;
  protected final boolean isFirstMove;
  private final int cachedHashCode;

  public Piece(final PieceType pieceType, final int piecePosition, final Alliance pieceAlliance,
      final boolean isFirstMove) {
    this.pieceType = pieceType;
    this.piecePosition = piecePosition;
    this.pieceAlliance = pieceAlliance;
    this.isFirstMove = isFirstMove;
    this.cachedHashCode = computeHashCode();
  }

  private int computeHashCode() {
    return this.cachedHashCode;
  }

  @Override
  public boolean equals(final Object other) {
    if (this == other) {
      return true;
    } else if (other instanceof Piece) {
      final Piece otherPiece = (Piece) other;
      return piecePosition == otherPiece.getPiecePosition() && pieceType == otherPiece.getPieceType()
          && pieceAlliance == otherPiece.getPieceAlliance() && isFirstMove == otherPiece.isFirstMove();
    }
    return false;
  }

  public int hashCode() {
    int result = pieceType.hashCode();
    result = 31 * result + pieceAlliance.hashCode();
    result = 31 * result + piecePosition;
    result = 31 * result + (isFirstMove ? 1 : 0);
    return result;
  }

  public Alliance getPieceAlliance() {
    return pieceAlliance;
  }

  public int getPiecePosition() {
    return piecePosition;
  }

  public PieceType getPieceType() {
    return pieceType;
  }

  public int getPieceValue() {
    return this.pieceType.getPieceValue();
  }

  public boolean isFirstMove() {
    return isFirstMove;
  }

  public abstract Collection<Move> calculateLegalMoves(final Board board);

  public abstract String toString();

  public abstract Piece movePiece(final Move move);

  public enum PieceType {
    PAWN("P", 100) {
      @Override
      public boolean isKing() {
        return false;
      }

      @Override
      public boolean isRook() {
        return false;
      }
    },
    KNIGHT("N", 300) {
      @Override
      public boolean isKing() {
        return false;
      }

      @Override
      public boolean isRook() {
        return false;
      }
    },
    BISHOP("B", 300) {
      @Override
      public boolean isKing() {
        return false;
      }

      @Override
      public boolean isRook() {
        return false;
      }
    },
    ROOK("R", 500) {
      @Override
      public boolean isKing() {
        return false;
      }

      @Override
      public boolean isRook() {
        return true;
      }
    },
    QUEEN("Q", 900) {
      @Override
      public boolean isKing() {
        return false;
      }

      @Override
      public boolean isRook() {
        return false;
      }
    },
    KING("K", 9999) {
      @Override
      public boolean isKing() {
        return true;
      }

      @Override
      public boolean isRook() {
        return false;
      }
    };

    private String pieceName;
    private int pieceValue;

    PieceType(final String pieceName, final int pieceValue) {
      this.pieceName = pieceName;
      this.pieceValue = pieceValue;
    }

    public int getPieceValue() {
      return this.pieceValue;
    }

    @Override
    public String toString() {
      return this.pieceName;
    }

    public abstract boolean isKing();

    public abstract boolean isRook();
  }
}
